targets/TARGET_STM/rtc_api.c

```diff
@@ -341,25 +346,51 @@ uint32_t rtc_read_lp(void)
 
 void rtc_set_wake_up_timer(timestamp_t timestamp)
 {
+    /* RTC periodic auto wake up timer is used
+    *  This WakeUpTimer is loaded to an init value => WakeUpCounter
+    *  then timer starts counting down (even in low-power modes)
+    *  When it reaches 0, the WUTF flag is set in the RTC_ISR register
+    *  The wakeup timer clock input is RTC clock divided by 2, 4, 8, or 16 => WakeUpClock
+    */
     uint32_t WakeUpCounter;
-    uint32_t current_lp_time;
+    uint32_t WakeUpClock = RTC_WAKEUPCLOCK_RTCCLK_DIV4;
 
-    current_lp_time = rtc_read_lp();
+    core_util_critical_section_enter();
 
+    /* MBED API gives the timestamp value to set
+    *  WakeUpCounter is then the delta between timestamp and the current tick (LPTICKER_counter)
+    *  If the current tick preceeds timestamp value, max U32 is added
+    */
+    uint32_t current_lp_time = rtc_read_lp();
     if (timestamp < current_lp_time) {
         WakeUpCounter = 0xFFFFFFFF - current_lp_time + timestamp;
     } else {
         WakeUpCounter = timestamp - current_lp_time;
     }
 
+    /* RTC WakeUpCounter is 16 bits
+    *  Corresponding time value depends on WakeUpClock
+    *  - RTC clock divided by 4  : max WakeUpCounter value is  8s
+    *  - RTC clock divided by 8  : max WakeUpCounter value is 16s
+    *  - RTC clock divided by 16 : max WakeUpCounter value is 32s
+    */
     if (WakeUpCounter > 0xFFFF) {
-        WakeUpCounter = 0xFFFF;
+        WakeUpClock = RTC_WAKEUPCLOCK_RTCCLK_DIV8;
+        WakeUpCounter = WakeUpCounter/2;
+
+        if (WakeUpCounter > 0xFFFF) {
+            WakeUpClock = RTC_WAKEUPCLOCK_RTCCLK_DIV16;
+            WakeUpCounter = WakeUpCounter/2;
+
+            if (WakeUpCounter > 0xFFFF) {
+                WakeUpCounter = 0xFFFF;
```

Comment:
In case where wake-up is pushed to the upper limit, it means that we cannot sleep as much as was requested right ? 
Is it then expected that higher layers (mbed-os core) takes care of synchronizing os-time again  ?

Reply:
Wake up can always be early, due to other wakeup sources than time. As long as the HAL reports how long it actually slept, that's fine, from the OS's point of view.

Reply:
The common ticker layer can safely handle early wakeups, but the LowPowerTickerWrapper does not. It treats early interrupts as spurious interrupts and ignores them.

To prevent this I recommend adjusting the supported `bits` from `lp_ticker_get_info` to match what you can safely set. This way the common ticker layer will always schedule a wakup time the hardware can achieve. In the case of maximum clock division (1/4th normal speed) with a 16 bit counter that would be `.bits=18`. This value is [currently set to 32 bits](https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_STM/lp_ticker.c#L253) which can cause failures.

Reply:
Isn't `.bits=18` still a little off? Can't actually schedule `0x3FFFF`, only `0x3FFFC`. Is that significant enough to matter?